include/openthread/tcp.h

```diff
@@ -0,0 +1,589 @@
+/*
+ *  Copyright (c) 2021, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * @brief
+ *  This file defines the OpenThread TCP API.
+ *
+ */
+
+#ifndef OPENTHREAD_TCP_H_
+#define OPENTHREAD_TCP_H_
+
+#include <openthread/ip6.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @addtogroup api-tcp
+ *
+ * @brief
+ *   This module includes functions that control TCP communication.
+ *
+ * @{
+ *
+ */
+
+/**
+ * This a linked buffer structure for use with TCP.
+ *
+ * A single @p otLinkedBuffer structure references an array of bytes in memory,
+ * via @p mData and @p mLength. The @p mNext field is used to form a chain of
+ * @p otLinkedBuffer structures.
+ */
+typedef struct otLinkedBuffer
+{
+    struct otLinkedBuffer *mNext;  ///< Pointer to the next linked buffer in this chain, or NULL if this is the end of the chain
+    const uint8_t* mData;          ///< Pointer to data referenced by this linked buffer
+    uint16_t mLength;              ///< Length of this linked buffer
+} otLinkedBuffer;
+
+struct otTcpActiveSocket;
+typedef struct otTcpActiveSocket otTcpActiveSocket;
+
+/**
+ * This callback informs the application that the TCP 3-way handshake is
+ * complete and that the connection is now established.
+ */
+typedef void (*otTcpConnectDone)(void *aContext, otTcpActiveSocket *aSocket);
```

Comment:
If the connection is never established, what event do we trigger for active sockets?

Reply:
You would get the `otTcpConnectionLost` callback if connection establishment fails.

Reply:
User may only expect `otTcpConnectionLost` after `otTcpConnectDone` was triggered. That make sense because without a established connection, how can we lose it?

Reply:
I think there's merit to that. How do you propose we change the API to clarify this? Do you think mentioning this in the doxygen is enough?

Reply:
Currently it is mentioned in the documentation, so I feel it's enough. If there's a particular change in the API that you'd like to see, suggest it here and we can discuss.

Reply:
Similar to #6063, we can have these events:
- otTcpConnectionEstablished --- triggered when TCP is connected (i.e. enter `ESTABLISHED` state)
- otTcpConnectionLost --- trigger when TCP is disconnected (i.e. enter `TIME-WAIT` state, or enter `CLOSED` state from states other than `TIME-WAIT` )
- otTcpClosed --- triggered when the TCP is closed (enters `CLOSED` state) for any reason

Since a TCP socket should enter `CLOSED` state eventually, `otTcpClosed` provides a convenient place for ALP to reuse or release the socket.  

Thoughts?

Reply:
To clarify, if a connection is in the ESTABLISHED state and then gets an RST packet, would it get both otTcpConnectionLost and otTcpClosed callbacks?

Reply:
> To clarify, if a connection is in the ESTABLISHED state and then gets an RST packet, would it get both otTcpConnectionLost and otTcpClosed callbacks?

Yes.

Reply:
If there’s broad support for this We can do this. I’ll give others a chance to chime in on how they feel about it.

I think it might be confusing to have two callbacks in response to a single event in some cases. If we go this route, we should define the order in which the callbacks are fired.

I understand that otTcpConnectionLost can be confusing in the case you described. I think that otTcpClosed can also be confusing in this API because we use the term “closed” to mean “uninitialized”. For example, after calling otTcpActiveShutdown, you don’t need to re-open the socket before using it.

How about otTcpDisconnected? I think that name might be the clearest. It has the property you said, that all TCP sockets would enter this state eventually. Thoughts?

Reply:
I changed the callbacks similarly to what was discussed above. Instead of `otTcpConnectDone` we now have `otTcpEstablished` that's called when a connection enters the established state. (So we can consider it "connected" from the moment `otTcpActiveConnect` is called on it.) Similary, I changed `otTcpConnectionLost` to `otTcpDisconnected`. In light of the previous change, I think this clarifies things, because now users aren't going to associate `otTcpDisconnected` with `otTcpEstablished` (since there's no `Connect` in `otTcpEstablished`).

Note that `otTcpDisconnected` is called when a connection enters time-wait and again when it is closed. It's similar to how you implemented callbacks in your slides, where you have a single callback function that gets called with various "event" flags. Although we can't do this for the other callbacks due to the arguments to those functions, we can do it here to avoid having two different callbacks where one will do.

I think this captures the best of both approaches. I just pushed a commit that does this.